tests/unittests/test_timeout.py

```diff
@@ -0,0 +1,244 @@
+            "refresh_token": "test_refresh_token",
+            "site_urls": "www.test.com",
+            "start_date": "2021-01-01T00:00:00Z",
+            "user_agent": "test_user_agent",
+            "request_timeout": 0.0
+        }
+
+        # initialize 'GoogleClient'
+        cl = client.GoogleClient(config['client_id'],
+                                 config['client_secret'],
+                                 config['refresh_token'],
+                                 config['site_urls'],
+                                 user_agent=config['user_agent'],
+                                 timeout_from_config=config.get('request_timeout'))
+
+        # verify that timeout value is default as request timeout is zero in the config
+        self.assertEquals(300, cl.request_timeout)
+
+    def test_timeout_string_0_value_passed_in_config(self):
+        config = {
+            "client_id": "test_client_id",
+            "client_secret": "test_client_secret",
+            "refresh_token": "test_refresh_token",
+            "site_urls": "www.test.com",
+            "start_date": "2021-01-01T00:00:00Z",
+            "user_agent": "test_user_agent",
+            "request_timeout": "0.0"
+        }
+
+        # initialize 'GoogleClient'
+        cl = client.GoogleClient(config['client_id'],
+                                 config['client_secret'],
+                                 config['refresh_token'],
+                                 config['site_urls'],
+                                 user_agent=config['user_agent'],
+                                 timeout_from_config=config.get('request_timeout'))
+
+        # verify that timeout value is default as request timeout is zero in the config
+        self.assertEquals(300, cl.request_timeout)
+
+@mock.patch("time.sleep")
+@mock.patch("requests.Session.request")
+class TestTimeoutBackoff(unittest.TestCase):
+    """
+        Verify that we backoff for 5 times for the 'Timeout' error
+    """
+
+    def test_timeout_error__get_access_token(self, mocked_request, mocked_sleep):
+
+        # mock request and raise the 'Timeout' error
+        mocked_request.side_effect = requests.Timeout
+
+        config = {
+            "client_id": "test_client_id",
+            "client_secret": "test_client_secret",
+            "refresh_token": "test_refresh_token",
+            "site_urls": "www.test.com",
+            "start_date": "2021-01-01T00:00:00Z",
+            "user_agent": "test_user_agent"
+        }
+
+        # initialize 'GoogleClient'
+        try:
+            with client.GoogleClient(config['client_id'],
+                                     config['client_secret'],
+                                     config['refresh_token'],
+                                     config['site_urls'],
+                                     user_agent=config['user_agent'],
+                                     timeout_from_config=config.get('request_timeout')) as cl:
+                pass
```

Comment:
Shouldn't this raise a different error to fail the test? We don't expect that the client should be able to instantiate. If we don't get the Timeout error we don't know for certain that the 5 requests actually correspond to the backoff that was implemented.

Reply:
Here the `get_access_token` function also uses timeout, and it is called from `request` and `__enter__` function, but when we place backoff on request then in the worst-case scenario the tap will backoff for 25 times (5 get_access_token and 5 request) when the timeout error is raised in get_access_token.
Hence we placed the timeout for get_access_token above __enter__ and request function.